New URL structure

martijngastkemper opened the pull request:
This is an exciting case for the image scaler: double extensions which are both an allowed output type: `foo.jpg.jpg`. This forces the original file (`foo.jpg`) to be converted to JPG. But the original file is `foo.jpg.jpg`. I don't think we have a way to determine the original file name. It could be both.

Let's discus.

martijngastkemper committed:
Update URL structure

- Add support for parameters other then width and height
- 500x500 becomes widht:500|height:500
- Add convert parameter to solve the issue with double extensions in the
  original file. (the reason this change has been made).
- Upgrade Jest
- Update README.md
- Add CHANGELOG.md

harmenjanssen commented:
Is this a real-world case? 
I don't see how the image scaler would be able to determine the original filename in this case, without explicitly looking up multiple options on S3. This would also be necessary when the original would be `foo.jpg.png`. 

One possible scenario could be to always look up the entire path to see if it exists, and if so, determine that the original file is requested without conversion. 
If not, chop off the last extension and try again. 

It will make all requests slower however... (it will also be harder to unit test because this logic is currently separate from looking up anything on S3, but that's another problem).

martijngastkemper commented:
I've put it together. Probably a lot to improve, but I think it's a good moment for a first review. Naming must be checked. And I've replaced the pipe with an underscore. The pipe caused an error somewhere in AWS, but the moment the pipe was in the URL the Lambda function isn't even executed.

I've a deploy version (MLK staging) for testing: https://7ij2wddzl7.execute-api.eu-west-1.amazonaws.com/staging/resize?key=/scaled/convert:webp_height:100/mock/event-1.jpg.webp

harmenjanssen commented on CHANGELOG.md:
```diff
@@ -0,0 +1,14 @@
+Some examples:
+
+- `/scaled/500x500/foobar.jpg` -> `/scaled/width:500_height:500/foobar.jpg`
+- `/scaled/x400/foobar.jpg.webp` -> `/scaled/height:400_convert:webp/foobar.jpg.webp`
```
I would spend one extra sentence here to describe the new format, and the fact that `webp` needs to be both in the options and as an extension.

harmenjanssen commented on README.md:
```diff
@@ -44,25 +42,25 @@ Subsequent requests to the bucket will therefore skip the Lambda function entire
+You can convert an image to a different format.
 
-- `/scaled/500x500/foobar.jpg.webp`: this will create a **webp** version of the file `foobar.jpg`, scaled to 500x500.
+- `/scaled/width:500_height:500_convert:webp/foobar.jpg.webp`: this will create a **webp** version of the file `foobar.jpg`, scaled to 500x500.
```
Same here: I think it's a good idea to put some extra emphasis on the fact that webp is both an option and an extension.

harmenjanssen commented on test/cleanupPath.test.js:
```diff
@@ -0,0 +1,14 @@
+const cleanupPath = require("../util/cleanupPath");
+
+describe("cleanup path", () => {
+  describe("Given a path with a leading path", () => {
```
```suggestion
  describe("Given a path with a leading slash", () => {
```

harmenjanssen commented on test/getImageMimeType.test.js:
```diff
@@ -0,0 +1,5 @@
+const getImageMimeType = require("../util/getImageMimeType.js");
+
+it("Strips the extension from the filename", () => {
```
I was confused by this descriptor. 
Maybe you can use something like `it("Uses the extension of the filename to determine mimetype")`

harmenjanssen replied:
(also, I would add a couple more cases, just to be complete and because it's cheap)

harmenjanssen commented on test/getInputFormat.test.js:
```diff
@@ -0,0 +1,7 @@
+it("Return the right input format", function () {
+  expect(getInputFormat("foo.jpg")).toBe("jpg");
+  expect(getInputFormat("foo.jpg.png")).toBe("png");
+  expect(getInputFormat("foo.jpg.png", "png")).toBe("jpg");
```
Hmm, this test makes me realize I find the signature of `getInputFormat` very difficult to read. This assertion was really surprising to me.
But I understand how it works when combined with the options string. I'll try to think of a way to make this API a little clearer.

martijngastkemper replied:
Did you come up with something? I tried to find another name, but keep on circling around "format" and "extension". They are both the same, but `getInputExtension` isn't better IMO. Or maybe something with "source" and "destination" to ditch the word "input" might do the trick.

harmenjanssen replied:
No, let's keep it like this. 
Maybe in the future I'll find some inspiration.

harmenjanssen commented on test/resizeImage.test.js:
```diff
@@ -69,48 +50,63 @@ describe("resizeImage", () => {
+    });
+  });
+
+  describe("Given a SVG image", () => {
```
```suggestion
  describe("Given an SVG image", () => {
```

harmenjanssen commented on util/getImageMimeType.js:
```diff
@@ -0,0 +1,13 @@
+    return "image/svg+xml";
+  }
+  return `image/${extension}`;
+};
```
This file has been renamed to camel-case, but note that `deflate-image.js` still has a dashed filename.

harmenjanssen commented on util/parseOptions.js:
```diff
@@ -0,0 +1,17 @@
+ * @returns {{}}
+ */
+module.exports = (optionsString) => {
+  let options = {};
```
This can be a `const`. `const` only protects against re-assignment, but the properties of the object may still be modified.

harmenjanssen commented on util/parseOptions.js:
```diff
@@ -0,0 +1,17 @@
+  let options = {};
+
+  optionsString.split("_").forEach((option) => {
+    const [action, ...params] = option.split(":");
+    // Currently only one parameter is supported.
+    options[action] = params[0];
```
You can unpack the first entry automatically, like this:

```suggestion
    const [action, param] = option.split(":");
    options[action] = param;
```

Any parameters that are not specified in the destructuring syntax won't be assigned.
Example:

```js
const [foo, bar] = "Foo bar baz".split(" "); // "baz" is simply ignored and not assigned.
```

harmenjanssen commented on util/resizeImage.js:
```diff
@@ -2,18 +2,51 @@ const SHARP = require("sharp");
+
+  const sharp = SHARP(body).withMetadata().resize(resizeOptions);
+
+  if (outputFormat) {
```
Is `outputFormat` ever undefined?

martijngastkemper replied:
It's a function argument.

harmenjanssen reviewed (commented):
Nice work! Looking good, nice to clean some things up. 

I'm having trouble reading the `getInputFormat` and `getOutputFormat` functions. Not because they're written unclearly, but because that stupid parameter just throws me off.

I'll try to think of a way to make this more explicit.

harmenjanssen reviewed (approved):
LGTM!

martijngastkemper commented:
@harmenjanssen Ik krijg die tests niet werkend op Ubuntu, lokaal wel. In de README staat iets over shard. Ik heb geprobeerd `npm_config_platform` en `npm_config_arch` aan te passen, maar dat veranderd bij mij lokaal niets. Weet jij hoe die fixtures kunnen worden geupdate of heb je een suggestie waar ik dat kan zoeken.

harmenjanssen commented:
Hmm, vaag.
Ik heb Sharp dus destijds toegevoegd aan de lock file dmv deze regel:

```
npm_config_platform=linux npm_config_arch=x64 yarn add sharp
```

Dit zodat de binaries gebuild worden als Linux x64 versies. 
Ik zie dat deze PR Sharp wel upgrade naar een nieuwe versie, maar is dat wel weer met een dergelijke one liner gebeurd? Anders zit daar misschien het probleem. En het kan sowieso natuurlijk wel zijn dat de nieuwe versie net een iets andere JPEG genereert en dat de fixtures misschien geupdate moeten worden?

martijngastkemper committed:
Run tests on MacOS

Image created on Linux aren't the same as on MacOS. So when creating
fixtures you have to do that on MacOS.

martijngastkemper commented:
I really don't understand how the test fixtures have ever worked cross-platform. I've fixed it by running the tests on MacOS runners. Developers working on Linux have a problem, but I invite them to come up with a better solution.

martijngastkemper commented:
@harmenjanssen Can you look at the last commit, changing to a MacOS runner.

harmenjanssen reviewed (approved):
I've thought about this, and came to the following conclusion: when I started this project I deployed from local to Lambda using Serverless. 
That meant my local binaries were shipped to AWS and quickly turned out to be incompatible.

That's when I found that one-liner that forces the Linux versions to be installed, even when on MacOS. (and apparently this was no problem when running this on MacOS)
This way I could ship the correct binaries to AWS from my local machine.

However, I think the upgrade in Sharp introduced an incompatibility. I also see you changed some of the fixtures in this PR. Not sure why that is, did you re-generate them?

In any case: I'm totally fine with running the tests on a MacOS runner. 
As long as we the tests are consistent between developer machines and CI, it's fine. 
As long as a working version is deployed to AWS, it's fine. 
It's too bad you can't deploy from your local machine, but that's not a hard requirement. (and maybe Serverless has a solution for this, come to think of it. Incompatible binaries are not uncommon in NPM land)

martijngastkemper commented:
> However, I think the upgrade in Sharp introduced an incompatibility. I also see you changed some of the fixtures in this PR. Not sure why that is, did you re-generate them?

Yes because my tests failed locally.

> It's too bad you can't deploy from your local machine, but that's not a hard requirement. (and maybe Serverless has a solution for this, come to think of it. Incompatible binaries are not uncommon in NPM land)

Changing `platfrom` and `arch` is the way to go: https://sharp.pixelplumbing.com/install#cross-platform
But this doesn't work for running code. Because it's not possible to run a macOS binary on Linux. So deploying from an incompatible OS is always tricky.

I'm going to merge this PR. It has taken way to much time. Maybe we find a better solution in the future.
